taichi/ir/ir.h

```diff
@@ -940,6 +941,23 @@ class WhileControlStmt : public Stmt {
   DEFINE_ACCEPT;
 };
 
+class ContinueStmt : public Stmt {
+ public:
+  // This is the loop on which this continue stmt has effects. It can be either
+  // an offloaded task, or a for/while loop inside the kernel.
+  Stmt *scope;
+
+  ContinueStmt() : scope(nullptr) {
+    TI_STMT_REG_FIELDS;
+  }
+
+  bool as_return() const {
```

Comment:
@yuanming-hu I just realized that this will require us to encapsulate the actual kernel body as a function (LLVM  does this, but Metal doesn't). Otherwise when we use one thread to handle a few elements, this could prematurely terminate the entire kernel thread.

Reply:
Using `continue` as a shortcut of `return` in top-level? Nice idea!

> Alternatively, do you folks think it even makes sense to support continue at the top level? I mean Taichi users are supposed to well understand that the top-level loop as parallelized, yet continue makes sense only for serial loops.

We want to keep this easy-to-use shortcut since **they acts same as the expected in python**.
In comparison, we don't want `break` in top-level is because **it can't act same as the expected in python**.

Reply:
> The other option is to directly translate this into a continue in the source backends. But we have to be very careful when doing that...

I guess this means we assume the kernel body is embedded inside a "meta" for-loop implemented by the Taichi runtime: https://github.com/taichi-dev/taichi/blob/58c64baf24b33898d217154d0026543e2aa0bc2e/taichi/runtime/llvm/runtime.cpp#L1074-L1077

Then I think wrapping the kernel into a function is a much more robust and cleaner approach...

Reply:
OK, sounds like I should keep this support. Thanks guys :)

Reply:
> Then I think wrapping the kernel into a function is a much more robust and cleaner approach...

Right, then I guess the source-to-source backends have to support something like non-linear code emission since now parallel loop bodies are outlined into functions. Pragmatically maybe using a `continue` is still the easiest way.

Reply:
> Right, then I guess the source-to-source backends have to support something like non-linear code emission since now parallel loop bodies are outlined into functions.

+1

> Pragmatically maybe using a continue is still the easiest way.

Right. I just realized a problem on Metal. Its `range_for` doesn't use grid-stride loop, yet its `struct_for` does, gahhh... 😦